oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java

```diff
@@ -376,15 +374,63 @@ public SampleFamily endpoint(List<String> serviceKeys, List<String> endpointKeys
         if (this == EMPTY) {
             return EMPTY;
         }
-        this.context.setMeterEntity(MeterEntity.newEndpoint(dim(serviceKeys), dim(endpointKeys)));
-        return left(io.vavr.collection.Stream.concat(serviceKeys, endpointKeys).asJava());
+        return createMeterSamples(serviceKeys, endpointKeys, ScopeType.ENDPOINT);
+    }
+
+    private SampleFamily createMeterSamples(List<String> serviceKeys, List<String> level2keys, ScopeType scopeType) {
+        final List<String> labelKeys = io.vavr.collection.Stream.concat(serviceKeys, level2keys).asJava();
+        Map<MeterEntity, Sample[]> meterSamples = new HashMap<>();
+        Arrays.stream(samples)
+              .collect(groupingBy(it -> getLabels(labelKeys, it), mapping(identity(), toList())))
+              .forEach((labels, samples) -> {
+                  MeterEntity meterEntity = getMeterEntity(samples, serviceKeys, level2keys, scopeType);
+                  Preconditions.checkNotNull(meterEntity);
+                  meterSamples.put(meterEntity, left(labelKeys, samples));
+              });
+
+        this.context.setMeterSamples(meterSamples);
+        //This samples is original, The grouped samples is in context which mapping with MeterEntity
+        return SampleFamily.build(this.context, samples);
+    }
+
+    private MeterEntity getMeterEntity(List<Sample> samples,
+                                       List<String> serviceKeys,
+                                       List<String> level2keys,
+                                       ScopeType scopeType) {
+        switch (scopeType) {
+            case SERVICE:
+                return MeterEntity.newService(dim(serviceKeys, samples));
+            case SERVICE_INSTANCE:
+                return MeterEntity.newServiceInstance(dim(serviceKeys, samples), dim(level2keys, samples));
+            case ENDPOINT:
+                return MeterEntity.newEndpoint(dim(serviceKeys, samples), dim(level2keys, samples));
+        }
+        return null;
     }
 
     private String dim(List<String> labelKeys) {
-        String name = labelKeys.stream().map(k -> samples[0].labels.getOrDefault(k, "")).collect(Collectors.joining("."));
+        String name = labelKeys.stream()
+                               .map(k -> samples[0].labels.getOrDefault(k, ""))
+                               .collect(Collectors.joining("."));
+        return CharMatcher.is('.').trimFrom(name);
+    }
+
+    private String dim(List<String> labelKeys, List<Sample> samples) {
+        String name = labelKeys.stream()
+                               .map(k -> samples.get(0).labels.getOrDefault(k, ""))
+                               .collect(Collectors.joining("."));
         return CharMatcher.is('.').trimFrom(name);
     }
 
+    private Sample[] left(List<String> labelKeys, List<Sample> samples) {
```

Comment:
@hanahmily These `dim` and `left` seem internal APIs, and as it takes `List<Sample> samples` as object, we should change these 2 methods as static. Please confirm.

@wankai123 `List<Sample> samples` should be the 1st parameter as the target.

Reply:
The `private` methods could not be accessed from expression, hence we don't have to convert them to static.

Reply:
Technically, yes. I asked for better class layout, because it becomes more and more complex.

Reply:
moved these codes to `InternalOps`